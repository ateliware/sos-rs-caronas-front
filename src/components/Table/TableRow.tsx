import { Fragment } from 'react';
import { FieldValues, UseFormWatch } from 'react-hook-form';

import { Icon } from '@components/Icon';

import { SelectedKeys } from './Table';
import TableCheckbox from './TableCheckbox';
import { ColumnProps } from './TableColumn';
import TableColumnCell from './TableColumnCell';

type Props = {
  row: any;
  checkable: boolean;
  dataKey: string;
  selectedKeys: SelectedKeys;
  invertSelection: boolean;
  tableId: string;
  columns: ColumnProps[];
  draggable?: boolean;
  setDragging?: (row: Props['row']) => void;
  onDrag?: (newIndex: number) => void;
  onDragEnd?: () => void;
  onSelectionChange?: (keys: SelectedKeys, invertSelection: boolean) => void;
  setSelectedKeys: (keys: SelectedKeys) => void;
  onClick?: (row: any) => void;
  watch: UseFormWatch<FieldValues>;
  showColumnSelector: boolean;
};

export default function TableRow(props: Props) {
  const rowClasses = ['table__row'];
  const dataKey = props.dataKey;
  const row = props.row;
  const selectedKeys = props.selectedKeys;
  const invertSelection = props.invertSelection;
  if (props.draggable) rowClasses.push('table__row--draggable');

  const dragOver = (e: any) => {
    e.preventDefault();
    const rows = Array.from(e.target.parentNode.parentNode.children);
    const row = e.target.parentNode;
    const newOrder = rows.indexOf(row);

    props?.onDrag?.(newOrder);
  };

  return (
    <tr
      className={rowClasses.join(' ')}
      onClick={(e) => {
        if ((e.target as HTMLElement).tagName !== 'DIV') props.onClick?.(row);
      }}
      draggable={props.draggable}
      onDragStart={() => props.setDragging?.(row)}
      onDragOver={props.draggable ? dragOver : undefined}
      onDragEnd={props.draggable ? () => props?.onDragEnd?.() : undefined}
    >
      {props.draggable && (
        <td className="table__cell">
          <div className="table__cell__content d-flex justify-center">
            <Icon>drag_indicator</Icon>
          </div>
        </td>
      )}
      {props.checkable && (
        <TableCheckbox
          checked={(selectedKeys[row[dataKey]] ?? false) !== invertSelection}
          onChange={(checked) => {
            const selected = checked !== invertSelection;
            const newSelectedKeys: SelectedKeys = {
              ...selectedKeys,
              [row[dataKey]]: true,
            };
            if (!selected) delete newSelectedKeys[row[dataKey]];

            props.setSelectedKeys(newSelectedKeys);
            props.onSelectionChange?.(newSelectedKeys, invertSelection);
          }}
        />
      )}
      {props.columns.map((column) => {
        const columnName = column.header?.toString() || '';

        if (
          props.showColumnSelector &&
          !props.watch(columnName, { value: true })
        )
          return <Fragment key={columnName}></Fragment>;

        return (
          <TableColumnCell
            key={props.tableId + column.header + row[dataKey]}
            {...column}
            row={row}
          />
        );
      })}
    </tr>
  );
}
