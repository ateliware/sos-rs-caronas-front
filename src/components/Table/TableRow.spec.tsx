import { fireEvent, render } from '@testing-library/react';

import TableRow from './TableRow';

describe('TableRow', () => {
  it('should render selecting', () => {
    const setSelectedKeys = jest.fn();
    const onSelectionChange = jest.fn();

    const { container } = render(
      <table>
        <tbody>
          <TableRow
            showColumnSelector={true}
            watch={jest.fn()}
            tableId="table"
            dataKey="id"
            row={{ id: 1, name: 'John Doe' }}
            columns={[
              { fromKey: 'id', header: 'Id' },
              { fromKey: 'name', header: 'Name' },
            ]}
            checkable
            selectedKeys={{}}
            invertSelection={false}
            setSelectedKeys={setSelectedKeys}
            onSelectionChange={onSelectionChange}
          />
        </tbody>
      </table>
    );
    expect(container).toMatchSnapshot();

    let checkbox = container.querySelector('.form-check-input') as Element;
    fireEvent.click(checkbox);
    expect(setSelectedKeys).toBeCalledWith({ '1': true });
    expect(onSelectionChange).toBeCalledWith({ '1': true }, false);
  });

  it('should render unselecting', () => {
    const setSelectedKeys = jest.fn();
    const onSelectionChange = jest.fn();

    const { container } = render(
      <table>
        <tbody>
          <TableRow
            showColumnSelector={true}
            watch={jest.fn()}
            tableId="table"
            dataKey="id"
            row={{ id: 1, name: 'John Doe' }}
            columns={[
              { fromKey: 'id', header: 'Id' },
              { fromKey: 'name', header: 'Name' },
            ]}
            checkable
            invertSelection
            selectedKeys={{ '1': true }}
            setSelectedKeys={setSelectedKeys}
            onSelectionChange={onSelectionChange}
          />
        </tbody>
      </table>
    );
    expect(container).toMatchSnapshot();

    let checkbox = container.querySelector('.form-check-input') as Element;
    fireEvent.click(checkbox);
    expect(setSelectedKeys).toBeCalledWith({});
    expect(onSelectionChange).toBeCalledWith({}, true);
  });
});
