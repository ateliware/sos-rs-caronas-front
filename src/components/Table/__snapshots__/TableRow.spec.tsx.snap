// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should render selecting 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="table__row"
      >
        <td
          class="table__cell table__cell--checkbox"
        >
          <div
            class="table__cell__content"
          >
            <div
              class="form-check"
            >
              <input
                class="form-check-input"
                id="checkbox"
                type="checkbox"
              />
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableRow should render unselecting 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="table__row"
      >
        <td
          class="table__cell table__cell--checkbox"
        >
          <div
            class="table__cell__content"
          >
            <div
              class="form-check"
            >
              <input
                class="form-check-input"
                id="checkbox"
                type="checkbox"
              />
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
